ClasspathResourceSelector is more lenient than ClasspathResourceSource

## Steps to reproduce

0. Have a test engine use JUnits `EngineDiscoveryRequestResolver`.
1. Discover with `DiscoverySelectors.selectClasspathResource("/")`.
2. Have the test engine call `.getClasspathResources()` on that selector. This will result in a `PreconditionViolationException`.
4. This exception is then caught by the `EngineDiscoveryRequestResolver` which tries to create an issue for it. 
5. This then fails because the `ClasspathResourceSource` does not accept the empty resource.

To reproduce clone: https://github.com/cucumber/cucumber-java-skeleton and replace the package selector in `RunCucumberTest` with `@SelectClasspathResource("/")`. Observe:


```
org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-platform-suite' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:208)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:174)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:119)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:84)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:104)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:91)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.junit.platform.launcher.core.InterceptingLauncher.lambda$execute$1(InterceptingLauncher.java:39)
	at org.junit.platform.launcher.core.ClasspathAlignmentCheckingLauncherInterceptor.intercept(ClasspathAlignmentCheckingLauncherInterceptor.java:25)
	at org.junit.platform.launcher.core.InterceptingLauncher.execute(InterceptingLauncher.java:38)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:66)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:231)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: org.junit.platform.commons.JUnitException: TestEngine with ID 'cucumber' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:208)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:174)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:119)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:102)
	at org.junit.platform.suite.engine.SuiteLauncher.discover(SuiteLauncher.java:58)
	at org.junit.platform.suite.engine.SuiteTestDescriptor.discover(SuiteTestDescriptor.java:126)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at org.junit.platform.suite.engine.DiscoverySelectorResolver.discoverSuites(DiscoverySelectorResolver.java:40)
	at org.junit.platform.suite.engine.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:48)
	at org.junit.platform.suite.engine.SuiteTestEngine.discover(SuiteTestEngine.java:60)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:195)
	... 17 more
Caused by: org.junit.platform.commons.PreconditionViolationException: Classpath resource name must not be null or blank
	at org.junit.platform.commons.util.Preconditions.condition(Preconditions.java:298)
	at org.junit.platform.commons.util.Preconditions.notBlank(Preconditions.java:266)
	at org.junit.platform.engine.support.descriptor.ClasspathResourceSource.<init>(ClasspathResourceSource.java:120)
	at org.junit.platform.engine.support.descriptor.ClasspathResourceSource.<init>(ClasspathResourceSource.java:116)
	at org.junit.platform.engine.support.descriptor.ClasspathResourceSource.from(ClasspathResourceSource.java:63)
	at org.junit.platform.launcher.core.DiscoveryIssueCollector.lambda$toSource$1(DiscoveryIssueCollector.java:95)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at org.junit.platform.launcher.core.DiscoveryIssueCollector.toSource(DiscoveryIssueCollector.java:95)
	at org.junit.platform.launcher.core.DiscoveryIssueCollector.selectorProcessed(DiscoveryIssueCollector.java:69)
	at org.junit.platform.launcher.listeners.discovery.CompositeLauncherDiscoveryListener.lambda$selectorProcessed$4(CompositeLauncherDiscoveryListener.java:61)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1092)
	at org.junit.platform.launcher.listeners.discovery.CompositeLauncherDiscoveryListener.selectorProcessed(CompositeLauncherDiscoveryListener.java:61)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:103)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.run(EngineDiscoveryRequestResolution.java:83)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolver.resolve(EngineDiscoveryRequestResolver.java:148)
	at io.cucumber.junit.platform.engine.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:32)
	at io.cucumber.junit.platform.engine.FeaturesPropertyResolver.resolveSelectors(FeaturesPropertyResolver.java:52)
	at io.cucumber.junit.platform.engine.CucumberTestEngine.discover(CucumberTestEngine.java:57)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:195)
	... 36 more
```

## Context

With JUnit 5.13.3.

Originally reported as https://github.com/cucumber/cucumber-jvm/issues/3024.

## Deliverables

- [ ] Make `ClasspathResourceSelector` reject the empty classpath resource name.
- [ ] Ensure that Suite Engine reports annotations with the empty resource name as a problem and doesn't throw when doing so.

--- a/documentation/src/docs/asciidoc/release-notes/release-notes-5.13.4.adoc
+++ b/documentation/src/docs/asciidoc/release-notes/release-notes-5.13.4.adoc
@@ -28,6 +28,8 @@ repository on GitHub.
 
 * `ClasspathResourceSelector` no longer allows to be constructed with a resource name that
   is blank after removing the leading slash.
+* `PackageSource.from(String)` now allows to be constructed with an empty string to
+  indicate the default package.
 
 [[release-notes-5.13.4-junit-platform-deprecations-and-breaking-changes]]
 ==== Deprecations and Breaking Changes

--- a/documentation/src/docs/asciidoc/release-notes/release-notes-5.13.4.adoc
+++ b/documentation/src/docs/asciidoc/release-notes/release-notes-5.13.4.adoc
@@ -26,7 +26,8 @@ repository on GitHub.
 [[release-notes-5.13.4-junit-platform-bug-fixes]]
 ==== Bug Fixes
 
-* ❓
+* `ClasspathResourceSelector` no longer allows to be constructed with a resource name that
+  is blank after removing the leading slash.
 
 [[release-notes-5.13.4-junit-platform-deprecations-and-breaking-changes]]
 ==== Deprecations and Breaking Changes

--- a/junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/ClasspathResourceSelector.java
+++ b/junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/ClasspathResourceSelector.java
@@ -25,6 +25,7 @@ import org.jspecify.annotations.Nullable;
 import org.junit.platform.commons.PreconditionViolationException;
 import org.junit.platform.commons.function.Try;
 import org.junit.platform.commons.support.Resource;
+import org.junit.platform.commons.util.Preconditions;
 import org.junit.platform.commons.util.ReflectionUtils;
 import org.junit.platform.commons.util.StringUtils;
 import org.junit.platform.commons.util.ToStringBuilder;
@@ -64,6 +65,8 @@ public final class ClasspathResourceSelector implements DiscoverySelector {
 	ClasspathResourceSelector(String classpathResourceName, @Nullable FilePosition position) {
 		boolean startsWithSlash = classpathResourceName.startsWith("/");
 		this.classpathResourceName = (startsWithSlash ? classpathResourceName.substring(1) : classpathResourceName);
+		Preconditions.notBlank(this.classpathResourceName,
+			"classpath resource name must not be blank after removing leading slash");
 		this.position = position;
 	}
 

--- a/junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/PackageSource.java
+++ b/junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/PackageSource.java
@@ -60,7 +60,10 @@ public final class PackageSource implements TestSource {
 	}
 
 	private PackageSource(String packageName) {
-		this.packageName = Preconditions.notBlank(packageName, "package name must not be null or blank");
+		Preconditions.notNull(packageName, "package name must not be null");
+		Preconditions.condition(packageName.isEmpty() || !packageName.isBlank(),
+			"package name must not contain only whitespace");
+		this.packageName = packageName;
 	}
 
 	/**

--- a/junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DiscoveryIssueCollector.java
+++ b/junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DiscoveryIssueCollector.java
@@ -10,6 +10,7 @@
 
 package org.junit.platform.launcher.core;
 
+import static org.junit.platform.commons.util.UnrecoverableExceptions.rethrowIfUnrecoverable;
 import static org.junit.platform.engine.SelectorResolutionResult.Status.FAILED;
 import static org.junit.platform.engine.SelectorResolutionResult.Status.UNRESOLVED;
 
@@ -78,7 +79,7 @@ class DiscoveryIssueCollector implements LauncherDiscoveryListener {
 		}
 	}
 
-	static @Nullable TestSource toSource(DiscoverySelector selector) {
+	private static @Nullable TestSource toSource(DiscoverySelector selector) {
 		if (selector instanceof ClassSelector classSelector) {
 			return ClassSource.from(classSelector.getClassName());
 		}
@@ -96,17 +97,26 @@ class DiscoveryIssueCollector implements LauncherDiscoveryListener {
 		if (selector instanceof PackageSelector packageSelector) {
 			return PackageSource.from(packageSelector.getPackageName());
 		}
-		if (selector instanceof FileSelector fileSelector) {
-			return fileSelector.getPosition() //
-					.map(DiscoveryIssueCollector::convert) //
-					.map(position -> FileSource.from(fileSelector.getFile(), position)) //
-					.orElseGet(() -> FileSource.from(fileSelector.getFile()));
+		try {
+			// Both FileSource and DirectorySource call File.getCanonicalFile() to normalize the reported file which
+			// can throw an exception for certain file names on certain file systems. UriSource.from(...) is affected
+			// as well because it may return a FileSource or DirectorySource
+			if (selector instanceof FileSelector fileSelector) {
+				return fileSelector.getPosition() //
+						.map(DiscoveryIssueCollector::convert) //
+						.map(position -> FileSource.from(fileSelector.getFile(), position)) //
+						.orElseGet(() -> FileSource.from(fileSelector.getFile()));
+			}
+			if (selector instanceof DirectorySelector directorySelector) {
+				return DirectorySource.from(directorySelector.getDirectory());
+			}
+			if (selector instanceof UriSelector uriSelector) {
+				return UriSource.from(uriSelector.getUri());
+			}
 		}
-		if (selector instanceof DirectorySelector directorySelector) {
-			return DirectorySource.from(directorySelector.getDirectory());
-		}
-		if (selector instanceof UriSelector uriSelector) {
-			return UriSource.from(uriSelector.getUri());
+		catch (Exception ex) {
+			rethrowIfUnrecoverable(ex);
+			logger.warn(ex, () -> "Failed to convert DiscoverySelector [%s] into TestSource".formatted(selector));
 		}
 		return null;
 	}

--- a/platform-tests/src/test/java/org/junit/platform/engine/discovery/DiscoverySelectorsTests.java
+++ b/platform-tests/src/test/java/org/junit/platform/engine/discovery/DiscoverySelectorsTests.java
@@ -299,8 +299,11 @@ class DiscoverySelectorsTests {
 		void selectClasspathResourcesPreconditions() {
 			assertViolatesPrecondition(() -> selectClasspathResource((String) null));
 			assertViolatesPrecondition(() -> selectClasspathResource(""));
+			assertViolatesPrecondition(() -> selectClasspathResource("/"));
 			assertViolatesPrecondition(() -> selectClasspathResource("    "));
+			assertViolatesPrecondition(() -> selectClasspathResource("/   "));
 			assertViolatesPrecondition(() -> selectClasspathResource("\t"));
+			assertViolatesPrecondition(() -> selectClasspathResource("/\t"));
 			assertViolatesPrecondition(() -> selectClasspathResource((Set<Resource>) null));
 			assertViolatesPrecondition(() -> selectClasspathResource(Collections.emptySet()));
 			assertViolatesPrecondition(() -> selectClasspathResource(Collections.singleton(null)));

--- a/platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/PackageSourceTests.java
+++ b/platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/PackageSourceTests.java
@@ -18,6 +18,8 @@ import java.io.Serializable;
 import java.util.stream.Stream;
 
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
 import org.junit.platform.commons.PreconditionViolationException;
 
 /**
@@ -49,9 +51,11 @@ class PackageSourceTests extends AbstractTestSourceTests {
 		assertThrows(PreconditionViolationException.class, () -> PackageSource.from((Package) null));
 	}
 
-	@Test
-	void packageSourceFromPackageName() {
-		var testPackage = getClass().getPackage().getName();
+	@ParameterizedTest
+	@ValueSource(classes = PackageSourceTests.class)
+	@ValueSource(strings = "DefaultPackageTestCase")
+	void packageSourceFromPackageName(Class<?> testClass) {
+		var testPackage = testClass.getPackage().getName();
 		var source = PackageSource.from(testPackage);
 
 		assertThat(source.getPackageName()).isEqualTo(testPackage);

--- a/platform-tests/src/test/java/org/junit/platform/launcher/core/DiscoveryIssueCollectorTests.java
+++ b/platform-tests/src/test/java/org/junit/platform/launcher/core/DiscoveryIssueCollectorTests.java
@@ -69,6 +69,7 @@ class DiscoveryIssueCollectorTests {
 			new Pair(selectClasspathResource("someResource", FilePosition.from(42, 23)),
 				ClasspathResourceSource.from("someResource",
 					org.junit.platform.engine.support.descriptor.FilePosition.from(42, 23))), //
+			new Pair(selectPackage(""), PackageSource.from("")), //
 			new Pair(selectPackage("some.package"), PackageSource.from("some.package")), //
 			new Pair(selectFile("someFile"), FileSource.from(new File("someFile"))), //
 			new Pair(selectFile("someFile", FilePosition.from(42)),

